Source/DafnyLanguageServer/Workspace/Notifications/VerificationDiagnosticsParams.cs

```diff
@@ -0,0 +1,640 @@
+      int numberOfLines,
+      int numberOfResolutionErrors,
+      bool verificationStarted,
+      Container<Diagnostic> diagnostics) {
+      var result = new LineVerificationStatus[numberOfLines];
+
+      if (verificationTrees.Length == 0 && numberOfResolutionErrors == 0 && verificationStarted) {
+        for (var line = 0; line < numberOfLines; line++) {
+          result[line] = LineVerificationStatus.Verified;
+        }
+
+        return result;
+      }
+
+      // Render verification tree content into lines.
+      foreach (var verificationTree in verificationTrees) {
+        if (verificationTree.Filename == verificationDiagnosticsParams.Uri.GetFileSystemPath() ||
+            "untitled:" + verificationTree.Filename == verificationDiagnosticsParams.Uri) {
+          verificationTree.RenderInto(result);
+        }
+      }
+
+      // Fill in the missing "Unknown" based on the surrounding content
+      // The filling only takes Verified an Error
+      var previousNotUnknown = LineVerificationStatus.Unknown;
+      var lineDelta = 1;
+      // Two passes so that we can fill gaps based on what happened before AND after
+      for (var line = 0; 0 <= line; line += lineDelta) {
+        if (line == numberOfLines) {
+          lineDelta = -1;
+          previousNotUnknown = LineVerificationStatus.Unknown;
+          continue;
+        }
+        if (previousNotUnknown != LineVerificationStatus.Verified &&
+            previousNotUnknown != LineVerificationStatus.VerifiedObsolete &&
+            previousNotUnknown != LineVerificationStatus.VerifiedVerifying) {
+          previousNotUnknown = LineVerificationStatus.Unknown;
+        }
+        if (result[line] == LineVerificationStatus.Unknown) {
+          result[line] = previousNotUnknown;
+        } else {
+          previousNotUnknown = result[line];
+        }
+      }
+
+      var resolutionErrorRendered = 0;
+      foreach (var diagnostic in diagnostics) {
+        if (resolutionErrorRendered >= numberOfResolutionErrors) {
+          break;
+        }
+        if (diagnostic.Range.Start.Line >= 0 && diagnostic.Range.Start.Line < result.Length) {
+          result[diagnostic.Range.Start.Line] = LineVerificationStatus.ResolutionError;
+          resolutionErrorRendered++;
+        }
+      }
+
+      var existsErrorRange = false;
+      var existsError = false;
+      foreach (var line in result) {
+        existsErrorRange = existsErrorRange || line == LineVerificationStatus.ErrorContext;
+        existsError = existsError || line == LineVerificationStatus.AssertionFailed;
+      }
+
+      if (existsErrorRange && !existsError) {
+        existsError = false;
+      }
+
+      return result;
+    }
+  }
+
+
+  public enum VerificationStatus {
+    Unknown = 0,
+    Verified = 200,
+    Inconclusive = 270,
+    Error = 400
+  }
+
+  public enum CurrentStatus {
+    Current = 0,
+    Obsolete = 1,
+    Verifying = 2
+  }
+
+  public enum LineVerificationStatus {
```

Comment:
I still have trouble understanding the meaning of these entries. The comments indicate that some values apply to assertions and other values apply to methods. Some values don't have a comment on them assigning them to other methods or assertions, so which do those belong to?

Does `Obsolete` mean, value is out of date and won't automatically become up to date, like when a user has 'verify on save' configured and has made edits?

Reply:
With comments from our call:

```
public enum LineVerificationStatus {
    // Default value for every line, before the renderer figures it out.
    // Used to separate lines that are not correctly verified.
    Nothing = 0,
    // For first-time computation not actively computing but soon. Synonym of "obsolete"
    // (scheduledComputation)
    // Either stale or queued. No previous result known.
    Scheduled = 1,
    // For first-time computations, actively computing
    Verifying = 2,

    // Context previously correctly verified, now stale or queued
    VerifiedObsolete = 201,
    // The context is previously correctly verified, and is not running.
    VerifiedVerifying = 202,

    // Either the context is verified, or there's an assertion in this line which verified and the context didn't error.
    // Or to context blank lines between contexts to make the green flow
    Verified = 200,

    // 300-302: For trees containing children with errors (e.g. functions, methods, fields, subset types)

    // The context previously had errors, but is now stale or queued.
    ErrorContextObsolete = 301,
    // The context previously had errors, but is now running.
    ErrorContextVerifying = 302,
    // The context has an error and no more work needs to be done.
    ErrorContextCompleted = 300,

    // 350-352: For individual assertions in error contexts

    // Note: take into account two assertions on one line.
    // Same as ErrorContextObsolete, but also the line contains no error diagnostics AND at least one assertion that passed in at least one batch.
    AssertionVerifiedInErrorContextObsolete = 351,
    // Same as ErrorContextVerifying, but also the line contains no error diagnostics AND at least one assertion that passed in at least one batch.
    AssertionVerifiedInErrorContextVerifying = 352,
    // Same as ErrorContextCompleted, but also the line contains no error diagnostics AND at least one assertion that passed in at least one batch.
    AssertionVerifiedInErrorContextCompleted = 350,
    // For specific lines which have errors on it. They take over verified assertions

    // Line previously contained at least one failed assertion, context is stale or queued.
    AssertionFailedObsolete = 401,
    // Line previously contained at least one failed assertion, context is running.
    AssertionFailedVerifying = 402,

    // Line contains at least one failed assertion, context can be in any state.
    AssertionFailed = 400,

    // For lines containing resolution or parse errors
    ResolutionError = 500
  }
```